src/streamlink/webbrowser/cdp/client.py

```diff
@@ -0,0 +1,406 @@
+    as well as executing JavaScript expressions and retrieving async results.
+
+    Don't instantiate this class yourself, use the :meth:`CDPClient.launch()` async context manager classmethod.
+    """
+
+    def __init__(self, cdp_connection: CDPConnection, nursery: trio.Nursery):
+        self.cdp_connection = cdp_connection
+        self.nursery = nursery
+
+    @classmethod
+    def launch(
+        cls,
+        session: Streamlink,
+        runner: Callable[[Self], Coroutine],
+        executable: Optional[str] = None,
+        timeout: Optional[float] = None,
+        cdp_host: Optional[str] = None,
+        cdp_port: Optional[int] = None,
+        cdp_timeout: Optional[float] = None,
+        headless: Optional[bool] = None,
+    ) -> Any:
+        """
+        Start a new :mod:`trio` runloop and do the following things:
+
+        1. Launch the Chromium-based web browser using the provided parameters or respective session options
+        2. Initialize a new :class:`CDPConnection` and connect to the browser's remote debugging interface
+        3. Create a new :class:`CDPClient` instance
+        4. Execute the async runner callback with the :class:`CDPClient` instance as only argument
+
+        If the ``webbrowser`` session option is set to ``False``, then a :exc:`CDPError` will be raised.
+
+        Example:
+
+        .. code-block:: python
+
+            async def my_app_logic(client: CDPClient):
+                async with client.session() as client_session:
+                    async with client_session.navigate("https://example.com") as frame_id:
+                        await client_session.loaded(frame_id)
+                        return await client_session.evaluate("document.body.querySelector('h1').innerText")
+
+            title: str | None = CDPClient.launch(session, my_app_logic)
+
+        :param session:     The Streamlink session object
+        :param runner:      An async client callback function which receives the :class:`CDPClient` instance as only parameter.
+        :param executable:  Optional path to the Chromium-based web browser executable.
+                            If unset, falls back to the ``webbrowser-executable`` session option.
+                            Otherwise, it'll be looked up according to the rules of the :class:`ChromiumBrowser` implementation.
+        :param timeout:     Optional global timeout value, including web browser launch time.
+                            If unset, falls back to the ``webbrowser-timeout`` session option.
+        :param cdp_host:    Optional remote debugging host.
+                            If unset, falls back to the ``webbrowser-cdp-host`` session option.
+                            Otherwise, ``127.0.0.1`` will be used.
+        :param cdp_port:    Optional remote debugging port.
+                            If unset, falls back to the ``webbrowser-cdp-port`` session option.
+                            Otherwise, a random free port will be chosen.
```

Comment:
> Otherwise, a random free port will be chosen.

Any concerns about random not including a specified range of upper level typically unused ports?

Reply:
Doesn't matter, because it's chosen by the system and not actually "random".
https://github.com/streamlink/streamlink/blob/20c8b6be542539681c96708d4f2f3f993463312b/src/streamlink/utils/socket.py#L6-L20